tools/pioasm/ada_output.cpp

```diff
@@ -0,0 +1,148 @@
+        std::string spec_file = package_name + ".ads";
+        FILE *out = fopen(spec_file.c_str(), "w");
+        if (!out) {
+            std::cerr << "Can't open output file '" << spec_file << "'" << std::endl;
+        }
+        return out;
+    }
+
+    void output_symbols(FILE *out, const std::vector<compiled_source::symbol> &symbols) {
+        int count = 0;
+        for (const auto &s : symbols) {
+            if (!s.is_label) {
+                fprintf(out, "%s : constant := %d;\n", s.name.c_str(), s.value);
+                count++;
+            }
+        }
+        if (count) {
+            fprintf(out, "\n");
+            count = 0;
+        }
+        for (const auto &s : symbols) {
+            if (s.is_label) {
+                fprintf(out, "   Offset_%s : constant := %d;\n", s.name.c_str(), s.value);
+                count++;
+            }
+        }
+        if (count) {
+            fprintf(out, "\n");
+        }
+    }
+
+    void ada_case(std::string &identifier) {
+        for(std::string::size_type i = 0; i < identifier.size(); ++i) {
+            if ((i == 0) || (identifier[i - 1] == '_')) {
+                identifier[i] = toupper(identifier[i]);
+            }
+        }
+    }
+
+    void header(FILE *out, const std::string msg, const int indent) {
+        const std::string dashes = std::string(msg.length() + 6, '-');
+        const std::string indent_str= std::string(indent, ' ');
+        fprintf(out, "%s%s\n", indent_str.c_str(), dashes.c_str());
+        fprintf(out, "%s-- %s --\n", indent_str.c_str(), msg.c_str());
+        fprintf(out, "%s%s\n", indent_str.c_str(), dashes.c_str());
+        fprintf(out, "\n");
+    }
+
+    int output(std::string destination, std::vector<std::string> output_options,
+               const compiled_source &source) override {
+
+        for (const auto &program : source.programs) {
+            for(const auto &p : program.lang_opts) {
+                if (p.first.size() >= name.size() && p.first.compare(0, name.size(), name) == 0) {
+                    std::cerr << "warning: " << name << " does not support output options; " << p.first << " lang_opt ignored.\n";
+                }
+            }
+        }
+
+        std::string package_name;
+
+        switch (output_options.size()) {
+            case 0:
+                std::cerr << "error: missing package name options for Ada format" << std::endl;
+                return 1;
+            case 1:
+                package_name = output_options[0]; // Package name from command options
+                break;
+            default:
+                std::cerr << "error: too many options for Ada format" << std::endl;
+                return 1;
+        }
+
+        FILE *out = open_single_output(destination);
+        if (!out) return 1;
+
+        header(out, "This file is autogenerated by pioasm; do not edit!", 0);
+        fprintf(out, "pragma Style_Checks (Off);\n\n");
+        fprintf(out, "with RP.PIO;\n\n");
```

Comment:
where would users get this from?

Reply:
https://github.com/JeremyGrosser/rp2040_hal/blob/master/src/drivers/rp-pio.ads

PIO example project here: https://github.com/JeremyGrosser/pico_examples/tree/master/pio_blink

We're still working out the details of all the APIs and build process, so the project is light on documentation at the moment. I expect that situation to improve soon.

Reply:
cool; i guess you are the first there, you can claim "ada', and therefore what it does is up to you... we don't need to have the supporting files, though perhaps a link somewhere would be helpful

Reply:
I put links to the relevant github repositories in a comment at the top of ada_output.cpp... Is that where it belongs or is there a better place to document this dependency?

Reply:
i think that is fine